dll/win32/fltlib/fltlib.c

```diff
@@ -56,15 +63,53 @@ FilterUnload(_In_ LPCWSTR lpFilterName)
 /* PRIVATE FUNCTIONS ****************************************************************/
 
 HRESULT
-NtStatusToHResult(_In_ NTSTATUS Status)
+FilterHResultFromNtStatus(_In_ NTSTATUS Status)
 {
-    HRESULT hr;
-    hr = RtlNtStatusToDosError(Status);
-    if (hr != ERROR_SUCCESS)
+    HRESULT Result;
+    ULONG DosError;
+
+    /* Check facility code of the status */
+    if (NT_FACILITY(Status))
     {
-        hr = (ULONG)hr | 0x80070000;
+        if (NT_FACILITY(Status) == FACILITY_FILTER_MANAGER)
+        {
+            /* Translate FilterManager specified status (STATUS_FLT_XXX) case-by-case */
+            if (Status == STATUS_FLT_BUFFER_TOO_SMALL)
+            {
+                Result = HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER);
+            }
+            else
+            {
+                /* Translate other FilterManager error codes by FILTER_HRESULT_FROM_FLT_NTSTATUS macro */
+                Result = FILTER_HRESULT_FROM_FLT_NTSTATUS(Status);
+            }
+        }
+        else
+        {
+            /* Facility is not FilterManager, return E_FAIL */
+            Result = E_FAIL;
+        }
     }
-    return hr;
+    else if (Status == STATUS_TIMEOUT)
+    {
+        Result = HRESULT_FROM_WIN32(WAIT_TIMEOUT);
+    }
+    else
+    {
+        /* Translate status to Win32 error */
+        DosError = RtlNtStatusToDosError(Status);
+
+        /* If RtlNtStatusToDosError didn't translate the status, we return the return value as-is */
+        if ((LONG)(DosError) <= 0)
+        {
+            return DosError;
+        }
+
+        /* Convert Win32 error to HRESULT */
+        Result = HRESULT_FROM_WIN32(DosError);
```

Comment:
Why translate it via dos to NT, instead of using `HRESULT_FROM_NT` ?
Are there apitests that show it needs to do the dance with the DosError?

Reply:
`HRESULT_FROM_NT(Status)` and `HRESULT_FROM_NT(RtlNtStatusToDosError(Status))` are totally different.

`HRESULT_FROM_NT` just wrap the Status in HResult format, no translation.
`FilterHResultFromNtStatus` returns HResult that contains Win32 error, not HResult that contains NTStatus.

And previous code did so.

Reply:
> HRESULT_FROM_NT(RtlNtStatusToDosError(Status))

You meant `HRESULT_FROM_WIN32(RtlNtStatusToDosError(Status))`

Reply:
> > HRESULT_FROM_NT(RtlNtStatusToDosError(Status))
> 
> You meant `HRESULT_FROM_WIN32(RtlNtStatusToDosError(Status))`

You're right, fixed